File: leetcode/medium/378_kth_smallest_element_in_a_sorted_matrix.py
# -----------------------------------------
# My Solution: Heap
#
# Time  Complexity: O(klog(n))
# Space Complexity: O(n)
# -----------------------------------------
import heapq

class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        row_size = col_size = len(matrix)

        # track elements by storing first row
        # for each iteration find the smallest elment, then add a new index (row_index + 1, col_index) to track_nums
        track_nums = []
        for col_index, cell_value in enumerate(matrix[0]):
            heapq.heappush(track_nums, (cell_value, 0, col_index))

        target_position = (0, 0)
        for _ in range(k)
            _, row_index, col_index = heapq.heappop(track_nums)
            if row_index + 1 < row_size:
                heapq.heappush(track_nums, (matrix[row_index + 1][col_index], row_index + 1, col_index))
            target_position = (row_index, col_index)
        return matrix[target_position[0]][target_position[1]]

# -----------------------------------------
# Model Solution: Heap
#
# Time  Complexity: O(klog(n))
# Space Complexity: O(n)
# -----------------------------------------
# Ref: https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/discuss/1248424/Python-O(klogk)
import heapq

class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        row_size = col_size = len(matrix)

        values_min_heap    = [(matrix[0][0], 0, 0)]
        visited_positions  = set([(0, 0)])
        kth_smallest_value = None
        for _ in range(k):
            kth_smallest_value, row_index, col_index = heapq.heappop(values_min_heap)
            for row_move, col_move in [(1, 0), (0, 1)]:
                next_row_index = row_index + row_move
                next_col_index = col_index + col_move
                if next_row_index == row_size or next_col_index == col_size or (next_row_index, next_col_index) in visited_positions:
                    continue
                heapq.heappush(values_min_heap, (matrix[next_row_index][next_col_index], next_row_index, next_col_index))
                visited_positions.add((next_row_index, next_col_index))

        return kth_smallest_value

File: leetcode/medium/378_kth_smallest_element_in_a_sorted_matrix.cpp
// -----------------------------------------
// My Solution: Heap
//
// Time  Complexity: O(mn)
// Space Complexity: O(k)
// -----------------------------------------
// m := matrix.size(), n := matrix[0].size()
class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        priority_queue<int> max_heap;
        for (int row = 0; row < matrix.size(); ++row) {
            for (int col = 0; col < matrix[0].size(); ++col) {
                max_heap.push(matrix[row][col]);
                if (max_heap.size() > k)
                    max_heap.pop();
            }
        }

        return max_heap.top();
    }
};

// -----------------------------------------
// Model Solution: Binary Search
//
// Time  Complexity: O((m + n) * log(k))
// Space Complexity: O(1)
// -----------------------------------------
// m := matrix.size(), n := matrix[0].size()
//
// Ref: https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/discuss/1322101/C%2B%2BPython-Binary-Search-Picture-Explain-Clean-and-Concise
class Solution {
public:
    int m, n;

    int kthSmallest(vector<vector<int>>& matrix, int k) {
        m = matrix.size(), n = matrix[0].size(); // For general, the matrix need not be a square
        int left = matrix[0][0], right = matrix[m-1][n-1], ans = -1;
        while (left <= right) {
            int mid = (left + right) >> 1;
            if (count_less_or_equal(matrix, mid) >= k) {
                ans = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return ans;
    }

    int count_less_or_equal(vector<vector<int>>& matrix, int x) {
        int cnt = 0, c = n - 1; // start with the rightmost column
        for (int r = 0; r < m; ++r) {
            while (c >= 0 && matrix[r][c] > x)
                --c;  // decrease column until matrix[r][c] <= x
            cnt += (c + 1);
        }
        return cnt;
    }
};
